ingester/ingester.go

```diff
@@ -0,0 +1,699 @@
+	for _, sample := range samples {
+		if err := i.append(ctx, sample); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (i *Ingester) append(ctx context.Context, sample *model.Sample) error {
+	for ln, lv := range sample.Metric {
+		if len(lv) == 0 {
+			delete(sample.Metric, ln)
+		}
+	}
+
+	i.stopLock.RLock()
+	defer i.stopLock.RUnlock()
+	if i.stopped {
+		return fmt.Errorf("ingester stopping")
+	}
+
+	state, err := i.getStateFor(ctx)
+	if err != nil {
+		return err
+	}
+
+	fp, series, err := state.getOrCreateSeries(sample.Metric)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		state.fpLocker.Unlock(fp)
+	}()
+
+	if sample.Timestamp == series.lastTime {
+		// Don't report "no-op appends", i.e. where timestamp and sample
+		// value are the same as for the last append, as they are a
+		// common occurrence when using client-side timestamps
+		// (e.g. Pushgateway or federation).
+		if sample.Timestamp == series.lastTime &&
+			series.lastSampleValueSet &&
+			sample.Value.Equal(series.lastSampleValue) {
+			return nil
+		}
+		i.discardedSamples.WithLabelValues(duplicateSample).Inc()
+		return ErrDuplicateSampleForTimestamp // Caused by the caller.
+	}
+	if sample.Timestamp < series.lastTime {
+		i.discardedSamples.WithLabelValues(outOfOrderTimestamp).Inc()
+		return ErrOutOfOrderSample // Caused by the caller.
+	}
+	prevNumChunks := len(series.chunkDescs)
+	_, err = series.add(model.SamplePair{
+		Value:     sample.Value,
+		Timestamp: sample.Timestamp,
+	})
+	i.memoryChunks.Add(float64(len(series.chunkDescs) - prevNumChunks))
+
+	if err == nil {
+		// TODO: Track append failures too (unlikely to happen).
+		i.ingestedSamples.Inc()
+	}
+	return err
+}
+
+func (u *userState) getOrCreateSeries(metric model.Metric) (model.Fingerprint, *memorySeries, error) {
+	rawFP := metric.FastFingerprint()
+	u.fpLocker.Lock(rawFP)
+	fp := u.mapper.mapFP(rawFP, metric)
+	if fp != rawFP {
+		u.fpLocker.Unlock(rawFP)
+		u.fpLocker.Lock(fp)
+	}
+
+	series, ok := u.fpToSeries.get(fp)
+	if ok {
+		return fp, series, nil
+	}
+
+	var err error
+	series, err = newMemorySeries(metric, nil, time.Time{})
+	if err != nil {
+		// err should always be nil when chunkDescs are nil
+		panic(err)
+	}
+	u.fpToSeries.put(fp, series)
+	u.index.add(metric, fp)
+	return fp, series, nil
+}
+
+// Query implements prism.Querier.
+func (i *Ingester) Query(ctx context.Context, from, through model.Time, matchers ...*metric.LabelMatcher) (model.Matrix, error) {
+	i.queries.Inc()
+
+	state, err := i.getStateFor(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	fps := state.index.lookup(matchers)
+
+	// fps is sorted, lock them in order to prevent deadlocks
+	queriedSamples := 0
+	result := model.Matrix{}
+	for _, fp := range fps {
+		state.fpLocker.Lock(fp)
+		series, ok := state.fpToSeries.get(fp)
+		if !ok {
+			state.fpLocker.Unlock(fp)
+			continue
+		}
+
+		values, err := samplesForRange(series, from, through)
+		state.fpLocker.Unlock(fp)
```

Comment:
I wonder if it's worth extracting this logic into a function just so the `Unlock` can be put in a `defer`.